File: Ejercicio 12/EJercicio 12.java
import java.util.*;

public class BucketSort {
    public static void bucketSort(float[] arr, int n) {
        @SuppressWarnings("unchecked")
        List<Float>[] buckets = new ArrayList[n];
        
        
        for (int i = 0; i < n; i++) {
            buckets[i] = new ArrayList<>();
        }
        
        for (int i = 0; i < n; i++) {
            int bucketIndex = (int)(n * arr[i]);
            if (bucketIndex == n) bucketIndex = n - 1;
            buckets[bucketIndex].add(arr[i]);
        }
        
        for (int i = 0; i < n; i++) {
            Collections.sort(buckets[i]);
        }
        
        int index = 0;
        for (int i = 0; i < n; i++) {
            for (float value : buckets[i]) {
                arr[index++] = value;
            }
        }
    }

    // Generar número aleatorio entre 0 y 1
    public static float generarAleatorio(Random random) {
        return random.nextFloat();
    }

    // Imprimir el arreglo
    public static void imprimirArray(float[] arr, int n) {
        for (int i = 0; i < n; i++) {
            System.out.printf("%.3f ", arr[i]);
        }
        System.out.println();
    }

    public static void main(String[] args) {
        Random random = new Random();

        final int TAMANO = 15;
        float[] arr = new float[TAMANO];

        // Generar 15 números aleatorios entre 0 y 1
        System.out.println("Generando 15 numeros aleatorios entre 0 y 1");
        for (int i = 0; i < TAMANO; i++) {
            arr[i] = generarAleatorio(random);
        }

        // Mostrar array original
        System.out.println("\nArray original:");
        imprimirArray(arr, TAMANO);

        // Aplicar Bucket Sort
        bucketSort(arr, TAMANO);

        System.out.println("\nArray ordenado con Bucket Sort:");
        imprimirArray(arr, TAMANO);
    }
}

File: Ejercicio 12/EJercicio 12.js
function bucketSort(arr, n) {
    let buckets = new Array(n);

    for (let i = 0; i < n; i++) {
        buckets[i] = [];
    }
    
    for (let i = 0; i < n; i++) {
        let bucketIndex = Math.floor(n * arr[i]);
        if (bucketIndex === n) bucketIndex = n - 1;
        buckets[bucketIndex].push(arr[i]);
    }
    
    for (let i = 0; i < n; i++) {
        buckets[i].sort((a, b) => a - b);
    }
    
    let index = 0;
    for (let i = 0; i < n; i++) {
        for (let value of buckets[i]) {
            arr[index++] = value;
        }
    }
}

// Generar número aleatorio entre 0 y 1
function generarAleatorio() {
    return Math.random();
}

// Imprimir el arreglo
function imprimirArray(arr, n) {
    let result = "";
    for (let i = 0; i < n; i++) {
        result += arr[i].toFixed(3) + " ";
    }
    console.log(result);
}

function main() {
    const TAMANO = 15;
    let arr = new Array(TAMANO);

    // Generar 15 números aleatorios entre 0 y 1
    console.log("Generando 15 numeros aleatorios entre 0 y 1");
    for (let i = 0; i < TAMANO; i++) {
        arr[i] = generarAleatorio();
    }

    // Mostrar array original
    console.log("\nArray original:");
    imprimirArray(arr, TAMANO);

    // Aplicar Bucket Sort
    bucketSort(arr, TAMANO);

    // Mostrar array ordenado
    console.log("\nArray ordenado con Bucket Sort:");
    imprimirArray(arr, TAMANO);
}

main();